### PR TITLE
Add JSON -> POJO type conversion sample

### LINKED ISSUE #13: Create Sample that showcases contentType conversion from JSON -> POJO automatically
As a user, I would like to refer to a sample that does the `application/json` contentType to a POJO automatically.

### PR DESCRIPTION
**Pull request by ilayaperumalg**
This resolves #13

### CODE REVIEW

File: stream-listener-json/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Build the sample by executing:
+
+	>$ mvn clean package
+
+## Running the Sample
+
+To start the source module execute the following:
+
+	>$ java -jar target/spring-cloud-stream-sample-stream-listener-json-1.0.0.BUILD-SNAPSHOT-exec.jar
+
```
Comment (markpollack) [inline stream-listener-json/README.md]: explain what output should be seen.

File: stream-listener-json/src/main/java/demo/SampleSink.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	@StreamListener(Sink.SAMPLE)
+	public void receive(Bar barMessage) {
+		System.out.println("******************");
+		System.out.println("At the Sink");
+		System.out.println("******************");
+		System.out.println("Received transformed message value: " + barMessage.getValue() + " of type " + barMessage.getClass());
+		System.out.println("******************");
+	}
+
+	public interface Sink {
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSink.java]: Why use this custom sink interface that duplicates the standard `Sink` class.  This adds unnecessary complexity to the app.

File: stream-listener-json/src/main/java/demo/SampleSink.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * @author Ilayaperumal Gopinathan
+ */
+@EnableBinding(SampleSink.Sink.class)
+public class SampleSink {
+
+	// Sink application definition
+	@StreamListener(Sink.SAMPLE)
+	public void receive(Bar barMessage) {
+		System.out.println("******************");
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSink.java]: No need for lines with `*****`

File: stream-listener-json/src/main/java/demo/SampleSink.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	public interface Sink {
+		String SAMPLE = "sample-sink";
+
+		@Input(SAMPLE)
+		SubscribableChannel sampleSink();
+	}
+
+	public static class Bar {
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSink.java]: Make this a standard class in a separate file and rename to `SimpleMessage` or something a bit more meaningful that `Bar`.  Also add a `toString()`.

File: stream-listener-json/src/main/java/demo/SampleSink.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+@EnableBinding(SampleSink.Sink.class)
+public class SampleSink {
+
+	// Sink application definition
+	@StreamListener(Sink.SAMPLE)
+	public void receive(Bar barMessage) {
+		System.out.println("******************");
+		System.out.println("At the Sink");
+		System.out.println("******************");
+		System.out.println("Received transformed message value: " + barMessage.getValue() + " of type " + barMessage.getClass());
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSink.java]: `"Received instance of " + simpleMessage.getClass() + ", " + simpleMesage.toString()`

File: stream-listener-json/src/main/java/demo/SampleSource.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ */
+@EnableBinding(SampleSource.Source.class)
+public class SampleSource {
+
+	@Bean
+	@InboundChannelAdapter(value = Source.SAMPLE, poller = @Poller(fixedDelay = "1000", maxMessagesPerPoll = "1"))
+	public MessageSource<String> timerMessageSource() {
+		return new MessageSource<String>() {
+			public Message<String> receive() {
+				System.out.println("******************");
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSource.java]: No need for `****` or `at the source` printlns

File: stream-listener-json/src/main/java/demo/SampleSource.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+				System.out.println("At the Source");
+				System.out.println("******************");
+				String message = "{\"value\":\"test\"}";
+				System.out.println("Sending JSON message: " + message);
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSource.java]: only need this println

File: stream-listener-json/src/main/java/demo/SampleSource.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+				System.out.println("******************");
+				return MessageBuilder.withPayload(message).setHeader("contentType", "application/json").build();
+			}
+		};
+	}
+
+	public interface Source {
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSource.java]: Why create a new source that mimics the standard one.  Just use the standard one.

File: stream-listener-json/README.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+In this *Spring Cloud Stream* sample, the application shows how to use StreamListener support to enable message mapping and automatic type conversion.
+
+## Requirements
+
+To run this sample, you will need to have installed:
+
+* Java 8 or Above
+
+This example requires Redis to be running on localhost.
```
Comment (markpollack) [inline stream-listener-json/README.md]: comment that the shared destination is configured in application.yml

File: stream-listener-json/src/main/java/demo/SampleSource.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+
+/**
+ * @author Ilayaperumal Gopinathan
+ */
+@EnableBinding(SampleSource.Source.class)
+public class SampleSource {
+
+	@Bean
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSource.java]: Interesting that this needs to be a @Bean to get the type conversion.  More basic example don't require this annotation.  ??

File: stream-listener-json/src/main/java/demo/SampleSource.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	@InboundChannelAdapter(value = Source.SAMPLE, poller = @Poller(fixedDelay = "1000", maxMessagesPerPoll = "1"))
```
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSource.java]: don't need add the poller explicitly, a 1 second poller will be created by default

**Comment by markpollack** [comment]
Pushed a variant of this sample here - https://github.com/markpollack/spring-cloud-stream-samples/tree/ilayaperumalg-stream-listener-json

File: stream-listener-json/src/main/java/demo/SampleSink.java (hunk 2 above)
Comment (ilayaperumalg) [inline stream-listener-json/src/main/java/demo/SampleSink.java]: My intention was to showcase the way someone could create a interface with the custom Channel definitions. But I agree, this is not the scope of this sample and could confuse the user.

File: stream-listener-json/src/main/java/demo/SampleSource.java (hunk 8 above)
Comment (markpollack) [inline stream-listener-json/src/main/java/demo/SampleSource.java]: Maybe use spring.cloud.stream.bindings.output.contentType=application/json instead and simplify the whole method to get rid of MessageSource and just return string.

**Comment by ilayaperumalg** [comment]
@markpollack Updated the PR with your changes and few additional changes based on the feedback (contentType usage and doc updates)

**Comment by sabbyanandan** [comment]
Closing this due to inactivity.

**Pull request closed without merging**